ygot/pathstrings.go

```diff
@@ -0,0 +1,268 @@
+func StringToPath(path string, pathTypes ...PathType) (*gnmipb.Path, error) {
+	var errs util.Errors
+	if len(pathTypes) == 0 {
+		return nil, util.AppendErr(errs, errors.New("no path types specified"))
+	}
+
+	pmsg := &gnmipb.Path{}
+	for _, p := range pathTypes {
+		switch p {
+		case StructuredPath:
+			gp, err := StringToStructuredPath(path)
+			if err != nil {
+				errs = util.AppendErr(errs, fmt.Errorf("error building structured path: %v", err))
+				continue
+			}
+			pmsg.Elem = gp.Elem
+		case StringSlicePath:
+			gp, err := StringToStringSlicePath(path)
+			if err != nil {
+				errs = util.AppendErr(errs, fmt.Errorf("error building string slice path: %v", err))
+				continue
+			}
+			pmsg.Element = gp.Element
+		}
+	}
+
+	if errs != nil {
+		return nil, errs
+	}
+
+	return pmsg, nil
+}
+
+// StringToStringSlicePath takes a string representing a path, and converts it into a
+// gnmi.Path. For example, if the Path "/a/b[c=d]/e" is input, it is converted
+// to a gnmi.Path{Element: []string{"a", "b[c=d]", "e"}} which is returned.
+func StringToStringSlicePath(path string) (*gnmipb.Path, error) {
+	parts := pathStringToElements(path)
+	for _, p := range parts {
+		// Run through extractKV to ensure that the path is valid.
+		if _, _, err := extractKV(p); err != nil {
+			return nil, fmt.Errorf("error parsing path %s: %v", path, err)
+		}
+	}
+
+	return &gnmipb.Path{
+		Element: parts,
+	}, nil
+}
+
+// StringToStructuredPath takes a string representing a path, and converts it to
+// a gnmi.Path, using the PathElem element message that is defined in gNMI 0.4.0.
+func StringToStructuredPath(path string) (*gnmipb.Path, error) {
+	parts := pathStringToElements(path)
+
+	gpath := &gnmipb.Path{}
+	for _, p := range parts {
+		name, kv, err := extractKV(p)
+		if err != nil {
+			return nil, fmt.Errorf("error parsing path %s: %v", path, err)
+		}
+		gpath.Elem = append(gpath.Elem, &gnmipb.PathElem{
+			Name: name,
+			Key:  kv,
+		})
+	}
+	return gpath, nil
+}
+
+// pathStringToElements splits the string s, which represents a gNMI string
+// path into its constituent elements. It does not parse keys, which are left
+// unchanged within the path - but removes escape characters from element
+// names. The path returned omits any leading empty elements when splitting
+// on the / character.
+func pathStringToElements(s string) []string {
+	var parts []string
+	var buf bytes.Buffer
+
+	var inKey, inEscape bool
+
+	for _, ch := range s {
+		switch {
+		case ch == '[' && !inEscape:
+			inKey = true
+		case ch == ']' && !inEscape:
+			inKey = false
+		case ch == '\\' && !inEscape && !inKey:
+			inEscape = true
+			continue
+		case ch == '/' && !inEscape && !inKey:
+			parts = append(parts, buf.String())
+			buf.Reset()
+			continue
+		}
+
+		buf.WriteRune(ch)
+		inEscape = false
+	}
+
+	if buf.Len() != 0 {
+		parts = append(parts, buf.String())
+	}
+
+	if len(parts) > 0 && parts[0] == "" {
```

Comment:
Was reading over gnmi spec and it looks like no trailing / should be an error? Anyway, seems like outputs should somehow be different for with/without leading /.

Reply:
I think that `/foo/bar` is completely legal - since this would refer to a leaf. It doesn't make sense to me that `/foo/container/` and `/foo/container` should have different outputs, since in all gNMI operations they mean the same thing. Could you explain why they might/should?

Reply:
i was referring to the case of a/b vs /a/b. sorry for the confusion, first trailing should be leading in my comment.